File: apps/api/providers/workos/adapter.py
from __future__ import annotations

from typing import Any

import httpx

from core.logging import get_logger
from providers.protocol import AuthEvent, AuthSession, ProviderUser
from providers.workos.client import WorkOSClient, get_workos_client
from providers.workos.errors import normalize_workos_error
from providers.workos.types.auth import WorkosAuthSuccess, WorkosUser

logger = get_logger(__name__)

_AUTH_FLOW_CODES = frozenset(
    {
        "email_verification_required",
        "mfa_enrollment",
        "mfa_challenge",
        "organization_selection_required",
        "sso_required",
        "organization_authentication_methods_required",
        "authentication_method_not_allowed",
        "email_password_auth_disabled",
        "passkey_progressive_enrollment",
        "radar_challenge",
        "radar_sign_up_challenge",
    }
)


class WorkOSAuthProvider:
    """Implements AuthProvider over the raw WorkOSClient HTTP client.

    - Returns AuthSession for successful authentication.
    - Returns AuthEvent for auth-flow steps (email verification required, MFA, etc.).
    - Raises AppHTTPException for hard errors (wrong password, banned, timeout).
    """

    def __init__(self, client: WorkOSClient) -> None:
        self._client = client

    # ── Helpers ───────────────────────────────────────────────────────────────

    @staticmethod
    def _to_provider_user(wu: WorkosUser) -> ProviderUser:
        return ProviderUser(
            id=wu.id,
            email=wu.email,
            first_name=wu.first_name,
            last_name=wu.last_name,
            email_verified=wu.email_verified,
            avatar=wu.profile_picture_url,
        )

    @staticmethod
    def _parse_success(raw: dict[str, Any]) -> AuthSession:
        success = WorkosAuthSuccess.model_validate(raw)
        return AuthSession(
            access_token=success.access_token,
            refresh_token=success.refresh_token,
            user=WorkOSAuthProvider._to_provider_user(success.user),
            organization_id=success.organization_id,
        )

    @staticmethod
    def _extract_event(exc: httpx.HTTPStatusError) -> AuthEvent | None:
        """Return AuthEvent for auth-flow codes; None for hard errors."""
        try:
            body = exc.response.json()
        except Exception:
            body = {}
        code: str = body.get("code") or body.get("error") or ""
        if code not in _AUTH_FLOW_CODES:
            return None
        return AuthEvent(
            kind=code,
            email=body.get("email"),
            pending_token=body.get("pending_authentication_token"),
            organizations=body.get("organizations") or [],
            auth_factors=body.get("authentication_factors") or [],
            connection_ids=body.get("connection_ids") or [],
        )

    def _handle_http_error(self, exc: httpx.HTTPStatusError) -> AuthEvent:
        """Return AuthEvent or raise AppHTTPException — never returns None."""
        event = self._extract_event(exc)
        if event is not None:
            return event
        logger.warning("workos.auth.hard_error", status=exc.response.status_code)
        raise normalize_workos_error(exc) from exc

    # ── Auth operations ────────────────────────────────────────────────────────

    async def login(
        self,
        *,
        email: str,
        password: str,
        client_id: str,
        ip_address: str | None = None,
        user_agent: str | None = None,
    ) -> AuthSession | AuthEvent:
        try:
            raw = await self._client.authenticate_with_password(
                email=email,
                password=password,
                client_id=client_id,
                ip_address=ip_address,
                user_agent=user_agent,
            )
            return self._parse_success(raw)
        except httpx.HTTPStatusError as exc:
            return self._handle_http_error(exc)

    async def register(
        self,
        *,
        email: str,
        password: str,
        first_name: str | None = None,
        last_name: str | None = None,
        email_verified: bool = False,
        metadata: dict[str, Any] | None = None,
    ) -> ProviderUser:
        try:
            raw = await self._client.create_user(
                email=email,
                password=password,
                first_name=first_name,
                last_name=last_name,
                email_verified=email_verified,
                metadata=metadata,
            )
            return self._to_provider_user(WorkosUser.model_validate(raw))
        except httpx.HTTPStatusError as exc:
            raise normalize_workos_error(exc) from exc

    async def get_user_by_email(self, *, email: str) -> ProviderUser | None:
        try:
            users = await self._client.list_users(email=email)
            if not users:
                return None
            return self._to_provider_user(WorkosUser.model_validate(users[0]))
        except httpx.HTTPStatusError as exc:
            raise normalize_workos_error(exc) from exc

    async def delete_user(self, *, user_id: str) -> None:
        try:
            await self._client.delete_user(user_id)
        except httpx.HTTPStatusError as exc:
            raise normalize_workos_error(exc) from exc

    async def send_otp(self, *, email: str, client_id: str) -> dict[str, Any]:
        try:
            return await self._client.create_magic_auth(email=email, client_id=client_id)
        except httpx.HTTPStatusError as exc:
            raise normalize_workos_error(exc) from exc

    async def verify_otp(self, *, code: str, email: str, client_id: str) -> AuthSession | AuthEvent:
        try:
            raw = await self._client.authenticate_with_magic_auth(code=code, email=email, client_id=client_id)
            return self._parse_success(raw)
        except httpx.HTTPStatusError as exc:
            return self._handle_http_error(exc)

    async def send_recovery(self, *, email: str, client_id: str) -> None:
        try:
            await self._client.create_password_reset(email=email, client_id=client_id)
        except httpx.HTTPStatusError as exc:
            raise normalize_workos_error(exc) from exc

    async def reset_password(self, *, token: str, new_password: str) -> ProviderUser:
        try:
            raw = await self._client.reset_password(token=token, new_password=new_password)
            return self._to_provider_user(WorkosUser.model_validate(raw["user"]))
        except httpx.HTTPStatusError as exc:
            raise normalize_workos_error(exc) from exc

    async def verify_email(
        self,
        *,
        code: str,
        pending_authentication_token: str,
        client_id: str,
    ) -> AuthSession | AuthEvent:
        try:
            raw = await self._client.authenticate_with_email_verification(
                code=code,
                pending_authentication_token=pending_authentication_token,
                client_id=client_id,
            )
            return self._parse_success(raw)
        except httpx.HTTPStatusError as exc:
            return self._handle_http_error(exc)

    async def authenticate_with_code(
        self,
        *,
        code: str,
        client_id: str,
        code_verifier: str | None = None,
        invitation_token: str | None = None,
        ip_address: str | None = None,
        user_agent: str | None = None,
    ) -> AuthSession | AuthEvent:
        try:
            raw = await self._client.authenticate_with_code(
                code=code,
                client_id=client_id,
                code_verifier=code_verifier,
                invitation_token=invitation_token,
                ip_address=ip_address,
                user_agent=user_agent,
            )
            return self._parse_success(raw)
        except httpx.HTTPStatusError as exc:
            return self._handle_http_error(exc)

    async def refresh(
        self,
        *,
        refresh_token: str,
        client_id: str,
        organization_id: str | None = None,
    ) -> AuthSession | AuthEvent:
        try:
            raw = await self._client.authenticate_with_refresh_token(
                refresh_token=refresh_token,
                client_id=client_id,
                organization_id=organization_id,
            )
            return self._parse_success(raw)
        except httpx.HTTPStatusError as exc:
            return self._handle_http_error(exc)

    async def revoke_session(self, *, session_id: str) -> None:
        await self._client.revoke_session(session_id=session_id)

    # ── Organization operations ───────────────────────────────────────────────

    async def create_organization(
        self,
        *,
        name: str,
        external_id: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        try:
            return await self._client.create_organization(name=name, external_id=external_id, metadata=metadata)
        except httpx.HTTPStatusError as exc:
            raise normalize_workos_error(exc) from exc

    async def create_organization_membership(
        self,
        *,
        user_id: str,
        organization_id: str,
        role_slug: str | None = None,
    ) -> dict[str, Any]:
        try:
            return await self._client.create_organization_membership(
                user_id=user_id,
                organization_id=organization_id,
                role_slug=role_slug,
            )
        except httpx.HTTPStatusError as exc:
            raise normalize_workos_error(exc) from exc

    async def delete_organization(self, *, organization_id: str) -> None:
        try:
            await self._client.delete_organization(organization_id)
        except httpx.HTTPStatusError as exc:
            raise normalize_workos_error(exc) from exc

    # ── Feature flag operations ───────────────────────────────────────────────

    async def add_feature_target(self, *, slug: str, target_id: str) -> None:
        try:
            await self._client.add_feature_flag_target(slug=slug, target_id=target_id)
        except httpx.HTTPStatusError as exc:
            raise normalize_workos_error(exc) from exc

    async def remove_feature_target(self, *, slug: str, target_id: str) -> None:
        try:
            await self._client.remove_feature_flag_target(slug=slug, target_id=target_id)
        except httpx.HTTPStatusError as exc:
            raise normalize_workos_error(exc) from exc

    # ── OAuth / JWKS ─────────────────────────────────────────────────────────

    def get_authorization_url(
        self,
        *,
        client_id: str,
        redirect_uri: str,
        provider: str | None = None,
        screen_hint: str | None = None,
        login_hint: str | None = None,
        state: str | None = None,
    ) -> str:
        return self._client.get_authorization_url(
            client_id=client_id,
            redirect_uri=redirect_uri,
            provider=provider,
            screen_hint=screen_hint,
            login_hint=login_hint,
            state=state,
        )

    async def get_jwks(self, *, client_id: str) -> dict[str, Any]:
        return await self._client.get_jwks(client_id=client_id)


# ── Factory ───────────────────────────────────────────────────────────────────

_provider_cache: WorkOSAuthProvider | None = None


def get_auth_provider(settings: Any) -> WorkOSAuthProvider:
    global _provider_cache
    if _provider_cache is None:
        _provider_cache = WorkOSAuthProvider(get_workos_client(settings))
    return _provider_cache

File: apps/api/src/providers/workos/adapter.ts
/**
 * The WorkOS implementation of the provider-neutral auth contract.
 *
 * Ported from `providers/workos/adapter.py`. This is the layer that turns a raw
 * WorkOS response into one of the three outcomes in `providers/auth.ts`:
 *
 * - a parsed {@link AuthSession} on success;
 * - an {@link AuthEvent} when WorkOS reports a **flow step** — verify an email,
 *   answer MFA, choose an organization — which is not a failure;
 * - a thrown `AppHttpError` for a hard error.
 *
 * `AUTH_FLOW_CODES` is what separates the second case from the third, and it is
 * the reason the client's `postAuth` re-throws the raw error: the flow codes
 * arrive as 4xx responses whose bodies carry the token the next step needs.
 */

import { getLogger } from '@/platform/logger'
import type { AuthEvent, AuthSession, ProviderUser } from '@/providers/auth'

import { getWorkOsClient, WorkOsClient } from './client'
import {
  isWorkOsHttpError,
  normalizeWorkOsError,
  WorkOsHttpError,
} from './errors'

const log = getLogger('workos')

/**
 * WorkOS codes that mean "another step is required", not "this failed".
 *
 * Anything outside this set is a hard error. Adding a code here changes a 4xx
 * from an error into a flow step, so it is a deliberate decision, not a
 * catch-all.
 */
const AUTH_FLOW_CODES: ReadonlySet<string> = new Set([
  'email_verification_required',
  'mfa_enrollment',
  'mfa_challenge',
  'organization_selection_required',
  'sso_required',
  'organization_authentication_methods_required',
  'authentication_method_not_allowed',
  'email_password_auth_disabled',
  'passkey_progressive_enrollment',
  'radar_challenge',
  'radar_sign_up_challenge',
])

function str(value: unknown): string | null {
  return typeof value === 'string' ? value : null
}

function objectList(value: unknown): Record<string, unknown>[] {
  return Array.isArray(value) ? (value as Record<string, unknown>[]) : []
}

function stringList(value: unknown): string[] {
  return Array.isArray(value)
    ? value.filter((item): item is string => typeof item === 'string')
    : []
}

/**
 * Read a WorkOS user payload, accepting either casing.
 *
 * WorkOS returns snake_case on most endpoints and camelCase on some; the Python
 * model declared both as validation aliases, so both are accepted here.
 */
export function toProviderUser(raw: Record<string, unknown>): ProviderUser {
  const pick = (...keys: string[]): string | null => {
    for (const key of keys) {
      const value = str(raw[key])
      if (value) return value
    }
    return null
  }

  const metadata = raw['metadata']

  return {
    id: str(raw['id']) ?? '',
    email: str(raw['email']) ?? '',
    firstName: pick('first_name', 'firstName'),
    lastName: pick('last_name', 'lastName'),
    emailVerified: Boolean(
      raw['email_verified'] ?? raw['emailVerified'] ?? false
    ),
    avatar: pick('profile_picture_url', 'profilePictureUrl'),
    metadata:
      metadata !== null &&
      typeof metadata === 'object' &&
      !Array.isArray(metadata)
        ? (metadata as Record<string, string>)
        : {},
  }
}

function parseSuccess(raw: Record<string, unknown>): AuthSession {
  const user = raw['user']

  return {
    accessToken: str(raw['access_token']) ?? str(raw['accessToken']) ?? '',
    refreshToken: str(raw['refresh_token']) ?? str(raw['refreshToken']),
    user: toProviderUser(
      user !== null && typeof user === 'object'
        ? (user as Record<string, unknown>)
        : {}
    ),
    organizationId:
      str(raw['organization_id']) ?? str(raw['organizationId']) ?? null,
  }
}

/** An {@link AuthEvent} for a flow code, or null when this is a hard error. */
function extractEvent(error: WorkOsHttpError): AuthEvent | null {
  if (!AUTH_FLOW_CODES.has(error.code)) return null

  return {
    kind: error.code,
    email: str(error.body['email']),
    pendingToken: str(error.body['pending_authentication_token']),
    organizations: objectList(error.body['organizations']),
    authFactors: objectList(error.body['authentication_factors']),
    connectionIds: stringList(error.body['connection_ids']),
  }
}

export class WorkOsAuthProvider {
  private readonly client: WorkOsClient

  constructor(client: WorkOsClient) {
    this.client = client
  }

  /** Return an {@link AuthEvent}, or throw. Never returns null. */
  private handleAuthError(error: unknown): AuthEvent {
    if (!isWorkOsHttpError(error)) throw error

    const event = extractEvent(error)
    if (event) return event

    log.warn({ status: error.status }, 'workos.auth.hard_error')
    throw normalizeWorkOsError(error)
  }

  /** Normalize a raw WorkOS failure; anything else passes through unchanged. */
  private rethrow(error: unknown): never {
    if (isWorkOsHttpError(error)) throw normalizeWorkOsError(error)
    throw error
  }

  // --- authentication ---

  async login(params: {
    email: string
    password: string
    clientId: string
    ipAddress?: string | null
    userAgent?: string | null
  }): Promise<AuthSession | AuthEvent> {
    try {
      return parseSuccess(await this.client.authenticateWithPassword(params))
    } catch (error) {
      return this.handleAuthError(error)
    }
  }

  async register(params: {
    email: string
    password?: string | null
    firstName?: string | null
    lastName?: string | null
    emailVerified?: boolean
    metadata?: Record<string, unknown> | null
  }): Promise<ProviderUser> {
    try {
      return toProviderUser(await this.client.createUser(params))
    } catch (error) {
      this.rethrow(error)
    }
  }

  async getUserByEmail(email: string): Promise<ProviderUser | null> {
    try {
      const users = await this.client.listUsers(email)
      const first = users[0]

      return first ? toProviderUser(first) : null
    } catch (error) {
      this.rethrow(error)
    }
  }

  async deleteUser(userId: string): Promise<void> {
    try {
      await this.client.deleteUser(userId)
    } catch (error) {
      this.rethrow(error)
    }
  }

  async sendOtp(
    email: string,
    clientId: string
  ): Promise<Record<string, unknown>> {
    try {
      return await this.client.createMagicAuth(email, clientId)
    } catch (error) {
      this.rethrow(error)
    }
  }

  async verifyOtp(params: {
    code: string
    email: string
    clientId: string
  }): Promise<AuthSession | AuthEvent> {
    try {
      return parseSuccess(await this.client.authenticateWithMagicAuth(params))
    } catch (error) {
      return this.handleAuthError(error)
    }
  }

  async sendRecovery(email: string, clientId: string): Promise<void> {
    try {
      await this.client.createPasswordReset(email, clientId)
    } catch (error) {
      this.rethrow(error)
    }
  }

  async resetPassword(
    token: string,
    newPassword: string
  ): Promise<ProviderUser> {
    try {
      const raw = await this.client.resetPassword(token, newPassword)
      const user = raw['user']

      return toProviderUser(
        user !== null && typeof user === 'object'
          ? (user as Record<string, unknown>)
          : {}
      )
    } catch (error) {
      this.rethrow(error)
    }
  }

  /**
   * Resend the email-verification code.
   *
   * The Python wrapped this in `except httpx.HTTPStatusError: self.
   * _handle_http_error(exc)`, but the client's `_post` had already converted the
   * failure into an `AppHTTPException` — so that clause could never fire and the
   * normalized error always propagated. The dead branch is not reproduced here;
   * this endpoint goes through the normalizing path and throws, which is what
   * the Python actually did.
   */
  sendVerificationEmail(userId: string): Promise<Record<string, unknown>> {
    return this.client.sendVerificationEmail(userId)
  }

  async verifyEmail(params: {
    code: string
    pendingAuthenticationToken: string
    clientId: string
  }): Promise<AuthSession | AuthEvent> {
    try {
      return parseSuccess(
        await this.client.authenticateWithEmailVerification(params)
      )
    } catch (error) {
      return this.handleAuthError(error)
    }
  }

  async authenticateWithCode(params: {
    code: string
    clientId: string
    codeVerifier?: string | null
    invitationToken?: string | null
    ipAddress?: string | null
    userAgent?: string | null
  }): Promise<AuthSession | AuthEvent> {
    try {
      return parseSuccess(await this.client.authenticateWithCode(params))
    } catch (error) {
      return this.handleAuthError(error)
    }
  }

  async refresh(params: {
    refreshToken: string
    clientId: string
    organizationId?: string | null
  }): Promise<AuthSession | AuthEvent> {
    try {
      return parseSuccess(
        await this.client.authenticateWithRefreshToken(params)
      )
    } catch (error) {
      return this.handleAuthError(error)
    }
  }

  /** Already normalized by the client, so no extra handling here. */
  revokeSession(sessionId: string): Promise<void> {
    return this.client.revokeSession(sessionId)
  }

  // --- organizations ---

  async createOrganization(params: {
    name: string
    externalId?: string | null
    metadata?: Record<string, unknown> | null
  }): Promise<Record<string, unknown>> {
    try {
      return await this.client.createOrganization(params)
    } catch (error) {
      this.rethrow(error)
    }
  }

  async createOrganizationMembership(params: {
    userId: string
    organizationId: string
    roleSlug?: string | null
  }): Promise<Record<string, unknown>> {
    try {
      return await this.client.createOrganizationMembership(params)
    } catch (error) {
      this.rethrow(error)
    }
  }

  async deleteOrganization(organizationId: string): Promise<void> {
    try {
      await this.client.deleteOrganization(organizationId)
    } catch (error) {
      this.rethrow(error)
    }
  }

  async deleteOrganizationMembership(membershipId: string): Promise<void> {
    try {
      await this.client.deleteOrganizationMembership(membershipId)
    } catch (error) {
      this.rethrow(error)
    }
  }

  // --- directory listing, for reconciliation ---

  async listUsers(): Promise<Record<string, unknown>[]> {
    try {
      return await this.client.listAllUsers()
    } catch (error) {
      this.rethrow(error)
    }
  }

  async listOrganizations(): Promise<Record<string, unknown>[]> {
    try {
      return await this.client.listAllOrganizations()
    } catch (error) {
      this.rethrow(error)
    }
  }

  async listOrganizationMemberships(
    filters: { organizationId?: string | null; userId?: string | null } = {}
  ): Promise<Record<string, unknown>[]> {
    try {
      return await this.client.listAllOrganizationMemberships(filters)
    } catch (error) {
      this.rethrow(error)
    }
  }

  // --- feature flags ---

  async addFeatureTarget(slug: string, targetId: string): Promise<void> {
    try {
      await this.client.addFeatureFlagTarget(slug, targetId)
    } catch (error) {
      this.rethrow(error)
    }
  }

  async removeFeatureTarget(slug: string, targetId: string): Promise<void> {
    try {
      await this.client.removeFeatureFlagTarget(slug, targetId)
    } catch (error) {
      this.rethrow(error)
    }
  }

  // --- OAuth / JWKS ---

  getAuthorizationUrl(params: {
    clientId: string
    redirectUri: string
    provider?: string | null
    screenHint?: string | null
    loginHint?: string | null
    state?: string | null
  }): string {
    return this.client.getAuthorizationUrl(params)
  }

  getJwks(clientId: string): Promise<Record<string, unknown>> {
    return this.client.getJwks(clientId)
  }
}

let cachedProvider: WorkOsAuthProvider | null = null

export function getAuthProvider(settings: {
  workos: { apiKey: string }
}): WorkOsAuthProvider {
  cachedProvider ??= new WorkOsAuthProvider(getWorkOsClient(settings))

  return cachedProvider
}

/** Drop the cached provider. Tests reconfigure the API key between suites. */
export function resetAuthProviderCache(): void {
  cachedProvider = null
}
